pkg/pillar/cmd/nodeagent/nodeagent.go

```diff
@@ -0,0 +1,484 @@
+			cmp.Diff(updated, sane))
+		globalConfig = sane
+		ctx.GCInitialized = true
+	}
+	log.Infof("handleGlobalConfigModify(%s): done\n", key)
+}
+
+func handleGlobalConfigDelete(ctxArg interface{},
+	key string, statusArg interface{}) {
+
+	ctx := ctxArg.(*nodeagentContext)
+	if key != "global" {
+		log.Infof("handleGlobalConfigDelete: ignoring %s\n", key)
+		return
+	}
+	log.Infof("handleGlobalConfigDelete for %s\n", key)
+	debug, _ = agentlog.HandleGlobalConfig(ctx.subGlobalConfig, agentName,
+		debugOverride)
+	globalConfig = types.GlobalConfigDefaults
+	log.Infof("handleGlobalConfigDelete done for %s\n", key)
+}
+
+// handle zedagent status events, for cloud connectivity
+func handleZedAgentStatusModify(ctxArg interface{},
+	key string, statusArg interface{}) {
+	ctx := ctxArg.(*nodeagentContext)
+	status := cast.ZedAgentStatus(statusArg)
+	if status.Key() != key {
+		log.Errorf("zedagentStatus key mismatch %s vs %s; ignored %+v\n",
+			key, status.Key(), status)
+		return
+	}
+	updateLastConfigReceivedTime(ctx, status)
+	log.Debugf("handleZedAgentStatusModify(%s) done\n", key)
+}
+
+func handleZedAgentStatusDelete(ctxArg interface{}, key string,
+	statusArg interface{}) {
+	// do nothing
+	log.Infof("handleZedAgentStatusDelete(%s) done\n", key)
+}
+
+// baseos status event handlers
+func handleBaseOsStatusModify(ctxArg interface{},
+	key string, statusArg interface{}) {
+	status := cast.CastBaseOsStatus(statusArg)
+	if status.Key() != key {
+		log.Errorf("baseOsStatus key mismatch %s vs %s; ignored %+v\n",
+			key, status.Key(), status)
+		return
+	}
+	log.Infof("handleBaseOsStatusModify(%s) done\n", key)
+}
+
+func handleBaseOsStatusDelete(ctxArg interface{},
+	key string, statusArg interface{}) {
+	log.Infof("handleBaseOsStatusDelete(%s) done\n", key)
+}
+
+// zboot status event handlers
+func handleZbootStatusModify(ctxArg interface{},
+	key string, statusArg interface{}) {
+	ctx := ctxArg.(*nodeagentContext)
+	status := cast.ZbootStatus(statusArg)
+	if status.Key() != key {
+		log.Errorf("zbootStatus key mismatch %s vs %s; ignored %+v\n",
+			key, status.Key(), status)
+		return
+	}
+	if status.CurrentPartition && ctx.updateInprogress &&
+		status.PartitionState == "active" {
+		log.Infof("CurPart(%s) marked as active\n", status.PartitionLabel)
+		ctx.updateInprogress = false
+		publishNodeAgentStatus(ctx)
+	}
+	doZbootBaseOsInstallationComplete(ctx, key, status)
+	doZbootBaseOsTestValidationComplete(ctx, key, status)
+	log.Infof("handleZbootStatusModify(%s) done\n", key)
+}
+
+func handleZbootStatusDelete(ctxArg interface{},
+	key string, statusArg interface{}) {
+
+	ctx := ctxArg.(*nodeagentContext)
+	if status := lookupZbootStatus(ctx, key); status == nil {
+		log.Infof("handleZbootStatusDelete: unknown %s\n", key)
+		return
+	}
+	log.Infof("handleZbootStatusDelete(%s) done\n", key)
+}
+
+// monitor device led blink change events
+func handleLedBlinkConfigModify(ctxArg interface{},
```

Comment:
The LED blinking was never designed to be operationally robust; merely to give feedback to an installer during onboard, So you shouldn't use it to guage the state of the system.

Furthermore, the logic here seems to have more than 10 cases. I think this needs to be explained in a state transition diagram, because there should only be a few cases:
1. Special onboarding - wait forever if we get 404 errors; if we don't get 404 errors then we should presumably reboot after ResetIfCloudGoneTime amount of time (to handle a hung device driver). The latter is not something we have today.
2. If baseimage update in progress, then if we don't get a config in the Fallback time, then reboot.
3. If not baseimage update in progress, then if we don't get a config in the ResetIfCloudGoneTime, then reboot.
Plus the logic for testcomplete. 
But that is an order of magnitude simpler than all of the different if cases in this function.

Reply:
sure, removed LED blinker based handling.